### PR TITLE
Fix comparisons

### PR DESCRIPTION
**Pull request by sisuresh**
### What

State expiration fields were not taken into account.

### CODE REVIEW

File: soroban-env-host/src/host/declared_size.rs
Comment (sisuresh) [inline soroban-env-host/src/host/declared_size.rs:150]: @jayz22 is this fine? Or is there an issue with declaring ContractDataEntry and LedgerKeyContractData a declared size type when they contain ScVals?

Comment (jayz22) [reply soroban-env-host/src/host/declared_size.rs]: We've discussed offline, for types with deep structures, we need to manually implement `MeteredClone` and `Compare` that calls into the substructures for meter charging. Otherwise we will just be charging for a shallow clone/cmp and underestimate the object size. Here we do not need the `declared_size` for `ContractDataEntry`, since we are breaking apart its substructures. 

I've added a couple of commits to your branch (fixing this and some other loose ends)  https://github.com/sisuresh/rs-soroban-env/pull/6

**Review by jayz22 — APPROVED** [review]

**Comment by jayz22** [comment]
@sisuresh I think this one should be good to go? There are some merge conflicts that need to be resolved but otherwise looks good.

**Comment by sisuresh** [comment]
> @sisuresh I think this one should be good to go? There are some merge conflicts that need to be resolved but otherwise looks good.

Yeah just need approval from someone with write access. @graydon maybe you could take a look when you have some time?

**Review by graydon — APPROVED** [review]

**Pull request merged**
